lib/exchange/provider/thorchain_exchange.provider.dart

```diff
@@ -0,0 +1,234 @@
+    String formattedToAddress = request.toAddress.startsWith('bitcoincash:')
+        ? request.toAddress.replaceFirst('bitcoincash:', '')
+        : request.toAddress;
+
+    final formattedFromAmount = double.parse(request.fromAmount);
+
+    final params = {
+      'from_asset': _normalizeCurrency(request.fromCurrency),
+      'to_asset': _normalizeCurrency(request.toCurrency),
+      'amount': _doubleToThorChainString(formattedFromAmount),
+      'destination': formattedToAddress,
+      'affiliate': _affiliateName,
+      'affiliate_bps': _affiliateBps,
+      'refund_address':
+          isRefundAddressSupported.contains(request.fromCurrency) ? request.refundAddress : '',
+    };
+
+    final responseJSON = await _getSwapQuote(params);
+
+    final inputAddress = responseJSON['inbound_address'] as String?;
+    final memo = responseJSON['memo'] as String?;
+
+    return Trade(
+        id: '',
+        from: request.fromCurrency,
+        to: request.toCurrency,
+        provider: description,
+        inputAddress: inputAddress,
+        createdAt: DateTime.now(),
+        amount: request.fromAmount,
+        state: TradeState.notFound,
+        payoutAddress: request.toAddress,
+        memo: memo);
+  }
+
+  @override
+  Future<Trade> findTradeById({required String id}) async {
+    if (id.isEmpty) throw Exception('Trade id is empty');
+    final formattedId = id.startsWith('0x') ? id.substring(2) : id;
+    final uri = Uri.parse('$_baseURL$_txInfoPath$formattedId');
+    final response = await http.get(uri);
+
+    if (response.statusCode == 404) {
+      throw Exception('Trade not found for id: $formattedId');
+    } else if (response.statusCode != 200) {
+      throw Exception('Unexpected HTTP status: ${response.statusCode}');
+    }
+
+    final responseJSON = json.decode(response.body);
+    final Map<String, dynamic> stagesJson = responseJSON['stages'] as Map<String, dynamic>;
+
+    final inboundObservedStarted = stagesJson['inbound_observed']?['started'] as bool? ?? true;
+    if (!inboundObservedStarted) {
+      throw Exception('Trade has not started for id: $formattedId');
+    }
+
+    final currentState = _updateStateBasedOnStages(stagesJson) ?? TradeState.notFound;
+
+    final tx = responseJSON['tx'];
+    final String fromAddress = tx['from_address'] as String? ?? '';
+    final String toAddress = tx['to_address'] as String? ?? '';
+    final List<dynamic> coins = tx['coins'] as List<dynamic>;
+    final String? memo = tx['memo'] as String?;
+    final String toAsset = memo != null ? (memo.split(':')[1]).split('.')[0] : '';
+
+    final plannedOutTxs = responseJSON['planned_out_txs'] as List<dynamic>?;
+    final isRefund = plannedOutTxs?.any((tx) => tx['refund'] == true) ?? false;
+
+    return Trade(
+      id: id,
+      from: CryptoCurrency.fromString(tx['chain'] as String? ?? ''),
+      to: CryptoCurrency.fromString(toAsset),
```

Comment:
these will not work properly when we include ERC20 tokens, we will probably need to refactor how this function works

Reply:
fixed

Reply:
how is this fixed?
if the `toAsset` is `USDT` then this function would return `USDT (OMNI)` since it's the first USDT asset

Reply:
fixed